### PR TITLE
implemented usage of numpy memmap for spilling to disk

### PR DESCRIPTION
**Pull request by pp1336**
Please do not auto merge as, auto merge will probably remove cache_blocking

### CODE REVIEW

**Comment by mlange05** [comment]
Can we please reconcile this with master before review?

File: examples/benchmark.py [hunk 1]
```diff
@@ -81,7 +82,13 @@ def source(t, f0):
...
 data.set_shape(nt, 30)
 # A Forward propagation example
-jit_obj = AcousticWave2D_cg(model1, data, create_dm, nbpml=nbpml, t_order=2, s_order=2)
+
+# path at which numpy memmap file are stored
+cache_disk_path = "/tmp/devito_disk"
+if not os.path.exists(cache_disk_path):
+    os.makedirs(cache_disk_path)
+
+jit_obj = AcousticWave2D_cg(model1, data, create_dm, nbpml=nbpml, t_order=2, s_order=2, disk_path=cache_disk_path)
```
Comment (navjotk) [inline examples/benchmark.py]: @mlange05 These are the kind of parameters you wouldn't have to pass if you had a config file.

**Comment by pp1336** [comment]
I have done a manual merge with the master, and also simplified the implementation so the only changes should be interfaces.py and tti_examples.py. Please take a look, thank you.

File: examples/benchmark.py (hunk 1 above)
Comment (mlange05) [inline examples/benchmark.py]: Surely, setting an output filepath is something the user should do by simply passing this in as a variable, ie. `u = TimeData(..., filepath='path/to/my/file')`, and not via configuration files or env vars. What would otherwise happen if we have multiple time variables that spill to multiple files?

That being said, I believe we are missing the bigger picture here with this whole config discussion, which (at least to my mind) is quite simply that we have two layers:
* 1. Devito itself acts as a library-like layer which provides the tools (eg, `TimeData`, `Operator`, etc.) for developers to build applications that are based on solving PDEs elegantly by defining them symbolically and associating them with input data. This core part of devito should follow general library design principles, which is it should be highly configurable at run-time and not enforce any particular workflow on it's users (ie. "Why do you want me to drop out of my IPython shell so badly?"). 
* 2. The FWI/TTI examples on the other hand are just that: examples of one particular user application that uses the abstractions provided by devito. They should really live in the outer `inversion` repo and hopefully will be migrated there soon. Here all the domain-specific knowledge is encoded, including readers for all types of funky file formats. And for that layer I quite frankly don't care how you configure them and where you get the filepaths from, since all the application has to do is simply propagate that filepath to the right devito object (in our case probably by setting it on `u`).
* The one exception to the above points, and I believe maybe this is where the confusion started, are, to my mind, architecture-specific configurations, like compiler flags and settings. These should be configured through env vars (unless explicitly set in the application code), in order to provide platform-portability. That means I should be able to move a fully configured FWI-Forward operator code that I wrote on a Mac to a different platform (ie. some HPC cluster), simply by setting one or maybe two env flags to tell the JIT engine to use a specific compiler setting, ideally without touching the Python code.

In short, I am not opposed to using configure files or any other means of setting any application-level parameters, as long as they are confined to the application layer (in this case the specific `AcousticWave2D` class in examples). What I do not want is to have to write 20 lines into an obscure devito configuration before doing a live demonstration to a student showing how I can solve a simple diffusion PDE in only two or three lines in an IPython shell.

File: devito/interfaces.py [hunk 2]
```diff
@@ -1,9 +1,47 @@
...
 class DenseData(IndexedBase):
+
+    # hash function used for generating part of memmap file name
+    _hashing_function = sha1
+    # this is the directory where memmap files will be created
+    _disk_path = None
+    # holds the string name of all memmap file created
+    _memmap_file_list = []
+
+    ## functions for managing memmap files
```
Comment (mlange05) [inline devito/interfaces.py]: Can we please keep all feature/functionality functions below `__init__()`? To my mind `__init__`/`__new__` should always be the first functions I see of any class.

File: devito/interfaces.py [hunk 3]
```diff
@@ -1,9 +1,47 @@
...
+from random import randint
+import os
 
 
 class DenseData(IndexedBase):
+
+    # hash function used for generating part of memmap file name
+    _hashing_function = sha1
+    # this is the directory where memmap files will be created
+    _disk_path = None
```
Comment (mlange05) [inline devito/interfaces.py]: Not sure this should be a class attribute. It is conceivable that a user might want different output files in different directories, no?

File: devito/interfaces.py [hunk 4]
```diff
@@ -1,9 +1,47 @@
...
+    # holds the string name of all memmap file created
+    _memmap_file_list = []
+
+    ## functions for managing memmap files
+    # call this method to specify where you want memmap file to be created
+    @staticmethod
+    def set_disk_path(disk_path):
+        if not os.path.exists(disk_path):
+            print("the directory" + disk_path + " does not exit")
```
Comment (mlange05) [inline devito/interfaces.py]: No need to be so verbose about this. Creating path directories that don't exist is expected behaviour.

That being said, if we had a proper logging mechanism, this might be appropriate on a "debug" level, but otherwise it will become annoying to users quickly.

File: devito/interfaces.py [hunk 5]
```diff
@@ -1,9 +1,47 @@
...
+    @staticmethod
+    def get_memmap_file_list():
+        return DenseData._memmap_file_list
+
+    @staticmethod
+    def remove_memmap_file():
+        for f in DenseData._memmap_file_list:
+            try:
+                os.remove(f)
+                print("removed file: " + f)
```
Comment (mlange05) [inline devito/interfaces.py]: Again, no need for this level of verbosity, I think.

File: devito/interfaces.py [hunk 6]
```diff
@@ -18,8 +56,23 @@ def set_initializer(self, lambda_initializer):
         assert(callable(lambda_initializer))
         self.initializer = lambda_initializer
 
+    # function to generate a hased file name used for memmap file
+    def _generate_memmap_filenmae(self):
```
Comment (mlange05) [inline devito/interfaces.py]: Why do we not just let a user input a filename/filepath, with a sensible default like `"/tmp/devito-memmap/data_%s" % self.name`?

File: devito/interfaces.py [hunk 7]
```diff
@@ -18,8 +56,23 @@ def set_initializer(self, lambda_initializer):
...
-        self.pointer = aligned(np.zeros(self.shape, self.dtype, order='C'), alignment=64)
+        self._disk_path = DenseData._disk_path
+        if self._disk_path == None:
+            self.pointer = aligned(np.zeros(self.shape, self.dtype, order='C'), alignment=64)
+        else:
+            f = self._disk_path + "/data_" + self.name + "_" + self._generate_memmap_filenmae()
+            self.pointer = aligned(np.memmap(filename=f, dtype=self.dtype, mode='w+', shape=tuple(self.shape), order='C'), alignment=64)
+            DenseData._memmap_file_list.append(f)
+            print("memmap file written to: " + f)
+            print("if the script call DenseData.remove_memmap_file() at the end, this file will be deleted automatically. if you have interupted this script, please go to " + self._disk_path + " to delete this file.")
```
Comment (mlange05) [inline devito/interfaces.py]: As a user I really don't want to see this message every time I run with a memmap.

**Comment by mlange05** [comment]
Ok, following the flurry of recent merges this needs to be reconciled with master again. Sorry about that.

Also, docstrings are missing throughout, and it would be great if we all become more conscientious about in-code documentation.

File: devito/interfaces.py (hunk 6 above)
Comment (pp1336) [inline devito/interfaces.py]: Hi, thank you very much. I can implement the user input, but I believe that we cant have defaults of the form "/tmp/devito-memmap/data_%s" % self.name, because there can be multiple "u" or "v" generated, this will overwritten previous data, therefore we are using a hash. Regarding the user entered name, I think that a prefix might be more appropriate because it will be quite a hassle for the user to input a filename every time a file is to be created. What do you think?

File: devito/interfaces.py (hunk 5 above)
Comment (pp1336) [inline devito/interfaces.py]: I will remove this.

**Comment by pp1336** [comment]
Hi, regarding the docstring, I have not written these class, just add some functions into the classes, I was just following the coding style. I will just add docstring for these classes then. Thank you.

File: devito/interfaces.py (hunk 6 above)
Comment (mlange05) [inline devito/interfaces.py]: Ok, I'm not sure what you mean by multiple "u"? `u = TimeData(...)` will always create one object with only one data buffer that you might want to spill and therefore require only one filepath. In fact the new symbolics infrastructure will effectively cache the data objects on their name so `TimeData(name='u', ...)` will always point to the same object and over-write the previous. So I'm really not convinced why you want to hash the file name.

Regarding the "hassle" of providing file names: That is why I want the symbol/object name, which the user always has to provide, to be part of the default path, so that most users won't need to overwrite the default. I do, however, want the option to manually overwrite this, because certain HPC clusters for example have special dedicated partitions for fast file IO, so an application developer might insert his preferred way to override the default paths; whether that is done via a prefix or a full path is your call.

**Comment by mlange05** [comment]
Yes, please provide docstrings for the functions you add, we'll take care of improving the class documentation. Admittedly, we have up to now been pretty naughty regarding in-place documentation, but as our developer group and feature range grows we all need to improve on this, so as to make everyone's life easier going forward.

File: devito/interfaces.py (hunk 6 above)
Comment (pp1336) [inline devito/interfaces.py]: Well, I guess then I will just use the name of the object then since the names are unique. If I mentioned because I remember seeing somewhere the an object of the same name were created twice, so I thought the name is not unique.

**Comment by pp1336** [comment]
Hi, I have made changes according to previous discussion, please take a look, thank you.

**Comment by mlange05** [comment]
Ok, will do. Can you please reconcile with master to allow auto-merge?

**Comment by pp1336** [comment]
Hi, I have already merged master into this branch, should be fine with auto merge, thanks.

**Comment by mlange05** [comment]
Yet github tells me you are 40 commits behind master. Have you fetched your local master branch?

**Comment by pp1336** [comment]
I think that its probably because I used meld to merge manually, I did fetch the latest master, I will just do git merge on memmap and push a new commit

**Comment by pp1336** [comment]
Hi, I have merged with master.

**Comment by mlange05** [comment]
Hey, you merged into master again, without checks passing! Your changes should be reviewed again and a full travis run needs to pass before you can merge! Do I really need to lock the master?

**Comment by pp1336** [comment]
Hi, sorry, I will remove the commit

**Comment by mlange05** [comment]
Ok, note that you will need to open a new PR for this branch now, since github won't let you re-open already merged ones.

**Comment by pp1336** [comment]
Hi, I think that I did something wrong and removed too much, I should I revert it? Thanks.

**Comment by pp1336** [comment]
Its fine now, I figured out how to revert it. Sorry for that.

File: examples/benchmark.py (hunk 1 above)
Comment (navjotk) [inline examples/benchmark.py]: @mlange05 I agree with the general principle in your comment above. However, the current reality is that we have an increasingly large number of parameters being passed around in the code. Some being hard-coded, some being picked up from environment variables and some as python parameters. A majority of these are being given "intelligent" default values. However the default values are spread all across the codebase since we have no convention on where the default value needs to be specified. Imagine the devito codebase 10 times as big as it is now and imagine a devito "user" trying to find out where in our codebase the default value for a particular parameter is coming from. Wouldn't it be easier to have it one Singleton config class?
